### ISSUE TITLE
"Last 7 days" does not actually show last 7 days

### ISSUE THREAD

**Issue by theofficialgman** [shlinkio/shlink-web-component#475]
### shlink-web-client version

4.2.1

### How do you use shlink-web-client

https://app.shlink.io

### Current behavior

selecting "Last 7 days" at the current time (UTC 4:20 PM Monday, October 14, 2024 Coordinated Universal Time (UTC)) show 2024-10-07 to 2024-10-15. This has been a longstanding issue where the date seems to always show 1 day ahead of the selected time period. The same issue occurs on the other date periods (eg: last 30 days and last 90 days).

This issue isn't a regression as I believe it has always worked like this.

### Expected behavior

I expect the "Last X days" period to actually show the last X days (eg: for today that would be 2024-10-7 to 2024-10-14)

### Minimum steps to reproduce

Go to any url on app.shlink.io and view the visits by time and select any of the "Last X days" periods

**Comment by acelaya** [comment]
I was about to publish a new shlink-web-client release including this "fix", but then I realized I was now experiencing the exact opposite issue. The last day in the range is skipped, so it only shows visits up until "yesterday", not up until "today".

I did not experience that when I implemented the fix, but looks like the workaround there had a reason to exist.

I think the solution will involve changing the approach more "deeply". I'll re-open this for now.

**Comment by acelaya** [comment]
Ok, I found the actual problem. For future context:

When the dots to be displayed are calculated, we pick the smallest (oldest) and biggest (newest) dates in the list of visits, and calculate the difference in hours/days/weeks/months (depending on what is selected in the dropdown).

That returns the number of dots to display. Then we iterate over those dots starting with the smallest date, and add the duration corresponding to selected gap (1 hour, 1 day, 1 week or 1 month).

The problem is that the difference is calculated taking into consideration partial dates, meaning the difference is a decimal number that gets rounded. When I originally implemented this I probably tested with an interval that resulted in a number being rounded down, which caused the last day of the interval to be skipped.

I thought that was simply because when calculating the difference we were skipping the last point, so I simply had to add 1 to the result.

However, this caused the interval to end up having one extra day if the difference got rounded up, causing the issue originally reported here.

In order to fix this, and make it display the right value, I'm calculating the absolute number of days/weeks/months between the two dates (AKA calendar days/calendar weeks/calendar months).
